### ISSUE TITLE
Avoiding undefined behaviour in rust::Str/String on construction

### ISSUE THREAD

**Issue by roligugus** [dtolnay/cxx#343]
rust::Str and rust::String constructors accept pointers and eventually call into slice::from_raw_parts() through cxxbridge04$str$valid() and cxxbridge04$string$from() from what I can gather.

Providing a nullptr to slice::from_raw_parts() is undefined behaviour, afaik.

We should probably add nullptr handling in the cases where it can happen, i.e. ptr/len constructor and ptr constructor. The std::string constructor seems save as it returns a valid pointer in the empty string case, see https://en.cppreference.com/w/cpp/string/basic_string/data

**Comment by roligugus** [comment]
Providing a `nullptr` to the `const char*` constructor will likely result in a crash on most platforms as it unconditionally calls `strlen()`.

I guess there could be a discussion if a plain `const char*` constructor should be provided in the first place, or if clients should always call the ptr/len constructor.

**Comment by dtolnay** [comment]
I would just say nullptr is not a legal input to any of the following constructors. I don't necessarily see it as something we need to check.
https://github.com/dtolnay/cxx/blob/4396abfafc69d5fe65fb10a5393a001d700a5043/include/cxx.h#L62-L63
https://github.com/dtolnay/cxx/blob/4396abfafc69d5fe65fb10a5393a001d700a5043/include/cxx.h#L32-L33

**Comment by roligugus** [comment]
Sounds like a reasonable approach. Worth documenting?

**Comment by dtolnay** [comment]
I got a chance to look into how std::string treats these cases. As far as I have been able to tell, `std::string(nullptr)` and `std::string(nullptr, >0)` are always UB but `std::string(nullptr, 0)` is actually fine. The aim of the cxx builtin bindings has always been to expose an idiomatic API to the other language, so cutting down differences between our string's API and std::string's is a useful design guide, to the extent that std::string is idiomatic in C++.

For `rust::Str(nullptr, 0)` and `rust::String(nullptr, 0)` we'd be able to choose any of:

- UB (as currently)
- terminate
- valid empty string (by bumping ptr up by 1)

@adetaylor I'm curious what you would consider most appropriate for your use case.
